pkg/esxi/esxi.go

```diff
@@ -0,0 +1,177 @@
+		return nil, err
+	}
+	// Don't pass the device to ESXi. It doesn't need it.
+	return loadConfig(filepath.Join(mountPoint, "boot.cfg"), "", 0)
+}
+
+// LoadConfig loads an ESXi configuration from configFile.
+func LoadConfig(configFile string) (*boot.MultibootImage, error) {
+	return loadConfig(configFile, "", 0)
+}
+
+func loadConfig(configFile, device string, partition int) (*boot.MultibootImage, error) {
+	opts, err := parse(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse config at %s: %v", configFile, err)
+	}
+	if len(device) > 0 {
+		if err := opts.addUUID(device, partition); err != nil {
+			return nil, fmt.Errorf("cannot add boot uuid of %s: %v", device, err)
+		}
+	}
+
+	return &boot.MultibootImage{
+		Path:    opts.kernel,
+		Cmdline: opts.args,
+		Modules: opts.modules,
+	}, nil
+}
+
+const (
+	kernel  = "kernel"
+	args    = "kernelopt"
+	modules = "modules"
+
+	comment = '#'
+	sep     = "---"
+
+	uuidMagic = "VMWARE FAT16    "
+	uuidSize  = 32
+)
+
+type options struct {
+	kernel  string
+	args    string
+	modules []string
+}
+
+func getUUID(device string, partition int) (string, error) {
+	device = strings.TrimRight(device, "/")
+	blockSize, err := gpt.GetBlockSize(device)
+	if err != nil {
+		return "", err
+	}
+
+	f, err := os.Open(fmt.Sprintf("%s%d", device, partition))
+	if err != nil {
+		return "", err
+	}
+
+	// Boot uuid is stored in the second block of the disk
+	// in the following format:
+	//
+	// VMWARE FAT16    <uuid>
+	// <---128 bit----><128 bit>
+	data := make([]byte, uuidSize)
+	n, err := f.ReadAt(data, int64(blockSize))
+	if err != nil {
+		return "", err
+	}
+	if n != uuidSize {
+		return "", io.ErrUnexpectedEOF
+	}
+
+	if magic := string(data[:len(uuidMagic)]); magic != uuidMagic {
+		return "", fmt.Errorf("bad uuid magic %q", magic)
+	}
+
+	uuid := hex.EncodeToString(data[len(uuidMagic):])
+	return fmt.Sprintf("bootUUID=%s", uuid), nil
+}
+
+func (o *options) addUUID(device string, partition int) error {
+	uuid, err := getUUID(device, partition)
+	if err != nil {
+		return err
+	}
+	o.args += " " + uuid
+	return nil
+}
+
+func parse(configFile string) (options, error) {
+	dir := filepath.Dir(configFile)
+
+	f, err := os.Open(configFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	var opt options
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		line = strings.TrimSpace(line)
+
+		if len(line) == 0 || line[0] == comment {
+			continue
+		}
+
+		tokens := strings.SplitN(line, "=", 2)
+		if len(tokens) != 2 {
+			return opt, fmt.Errorf("bad line %q", line)
+		}
+		key := strings.TrimSpace(tokens[0])
+		val := strings.TrimSpace(tokens[1])
+		switch key {
+		case kernel:
+			opt.kernel = filepath.Join(dir, val)
+		case args:
+			opt.args = filepath.Join(dir, val)
+		case modules:
+			for _, tok := range strings.Split(val, sep) {
+				tok = strings.TrimSpace(tok)
+				entry := strings.Fields(tok)
+				entry[0] = filepath.Join(dir, entry[0])
```

Comment:
What is this line doing?

Reply:
is there an assumption that the first entry in each token is some kind of filename? It's not immediately clear to me what format this is trying to parse

Reply:
I'll comment and write a test for it, too